recursion/reading-recursive-code.md

```diff
@@ -0,0 +1,288 @@
+
+Read through this recursive function, `pow`, whose responsibility is to calculate [exponentiation](https://en.wikipedia.org/wiki/Exponentiation), otherwise known as `num`<sup>`exponent`</sup>. It takes in two arguments:
+
+1. `num`, which is the base number
+1. `exponent`, which is the exponent
+
+For example, 3<sup>5</sup>, where `3` is `num` and `5` is `exponent`, is expressed and calculated like this:
+
+```
+3 * 3 * 3 * 3 * 3
+```
+
+And 3<sup>6</sup> is expressed as:
+
+```
+3 * 3 * 3 * 3 * 3 * 3
+```
+
+```python
+def pow(num, exponent):
+    if exponent == 0:
+        return 1
+
+    return num * pow(num, exponent - 1)
+```
+
+#### Base Case
+
+In exponentiation, any number raised to the power of zero equals one.
+
+```python
+if exponent == 0:
+    return 1
+```
+
+This code snippet is the _base case_. Whenever `pow` is called **and** the argument `exponent` is `0`, then we return `1`. This base case gives us an immediate solution to the problem: `1`!
+
+#### Recursive Cases
+
+```python
+return num * pow(num, exponent - 1)
+```
+
+This code snippet handles the recursive cases. Recursive cases _call itself while making the problem smaller_. Recursive cases should move the problem closer and closer to the base case.
+
+The problem gets smaller by calling `pow` with a smaller exponent. As the `exponent` shrinks, it will eventually equal `0`!
+
+### Running `pow`
+
+How can we visualize the flow of code execution? Let's imagine the same `pow` function with several print statements, and then run it.
+
+```python
+def pow(num, exponent):
+    print("Calling pow! num:", num, "exponent:", exponent)
+    if exponent == 0:
+        return 1
+
+    return num * pow(num, exponent - 1)
+
+
+print(3, "raised to the power of", 5, "is", pow(3, 5))
+
+print("-------------")
+
+print(3, "raised to the power of", 6, "is", pow(3, 6))
+```
+
+We should get the following output:
+
+```
+Calling pow! num: 3 exponent: 5
+Calling pow! num: 3 exponent: 4
+Calling pow! num: 3 exponent: 3
+Calling pow! num: 3 exponent: 2
+Calling pow! num: 3 exponent: 1
+Calling pow! num: 3 exponent: 0
+3 raised to the power of 5 is 243
+-------------
+Calling pow! num: 3 exponent: 6
+Calling pow! num: 3 exponent: 5
+Calling pow! num: 3 exponent: 4
+Calling pow! num: 3 exponent: 3
+Calling pow! num: 3 exponent: 2
+Calling pow! num: 3 exponent: 1
+Calling pow! num: 3 exponent: 0
+3 raised to the power of 6 is 729
+```
+
+We can confirm that the `pow` function was called multiple times. Each time, the problem was made smaller, and `exponent` decremented by one. Eventually, when `exponent` was `0`, it met the conditions for the base case, and returned `1`.
+
+## Infinite Recursion
+
+A function that calls itself without working towards a base case will repeat infinitely.
+
+```python
+def infinite_recursion(n):
+    return n + infinite_recursion(n)
+```
+
+A function without a base case will never have a stopping point! Consider this code which has no base case:
```

Comment:
emphasize the difference between modifying the argument and having a base case
```suggestion
Even if the function modifies its argument, if it doesn't have a base case it will never stop! Consider this code which decrements its argument but has no base case:
```